pkg/common/scheduler/state/interface.go

```diff
@@ -0,0 +1,209 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"context"
+	"errors"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	PodFitsResources                   = "PodFitsResources"
+	NoMaxResourceCount                 = "NoMaxResourceCount"
+	EvenPodSpread                      = "EvenPodSpread"
+	AvailabilityNodePriority           = "AvailabilityNodePriority"
+	AvailabilityZonePriority           = "AvailabilityZonePriority"
+	LowestOrdinalPriority              = "LowestOrdinalPriority"
+	RemoveWithEvenPodSpreadPriority    = "RemoveWithEvenPodSpreadPriority"
+	RemoveWithAvailabilityNodePriority = "RemoveWithAvailabilityNodePriority"
+	RemoveWithAvailabilityZonePriority = "RemoveWithAvailabilityZonePriority"
+	RemoveWithHighestOrdinalPriority   = "RemoveWithHighestOrdinalPriority"
+)
+
+// Plugin is the parent type for all the scheduling framework plugins.
+type Plugin interface {
+	Name() string
+}
+
+type FilterPlugin interface {
+	Plugin
+	// Filter is called by the scheduler.
+	// All FilterPlugins should return "Success" to declare that
+	// the given pod fits the vreplica.
+	Filter(ctx context.Context, args interface{}, state *State, key types.NamespacedName, podID int32) *Status
+}
+
+// ScoreExtensions is an interface for Score extended functionality.
+type ScoreExtensions interface {
+	// NormalizeScore is called for all pod scores produced by the same plugin's "Score"
+	// method. A successful run of NormalizeScore will update the scores list and return
+	// a success status.
+	NormalizeScore(ctx context.Context, state *State, scores PodScoreList) *Status
+}
+
+type ScorePlugin interface {
+	Plugin
+	// Score is called by the scheduler.
+	// All ScorePlugins should return "Success" unless the args are invalid.
+	Score(ctx context.Context, args interface{}, state *State, feasiblePods []int32, key types.NamespacedName, podID int32) (uint64, *Status)
+
+	// ScoreExtensions returns a ScoreExtensions interface if it implements one, or nil if does not
+	ScoreExtensions() ScoreExtensions
+}
+
+// NoMaxResourceCountArgs holds arguments used to configure the NoMaxResourceCount plugin.
+type NoMaxResourceCountArgs struct {
```

Comment:
same comment as above, seems like it should belongs in the NoMaxResourceCount plugin